### PR TITLE
Cleanup, fixes, and tests

### PR DESCRIPTION
**Pull request by znado**
Still a few loose ends to wrap up but ready for a first review.

Refactoring evals to use `eval_model()` in the root abstract class in `spec.py`, and have all workloads define `_eval_model_on_split()`. This will also make it so that metrics returned by `eval_model()` have the dataset split prepended to them with a `/`, so for examples `eval_train/accuracy`.

Adding `tests/reference_submission_tests.py`. This is a brief test that runs the for the workload and reference submission code for one train and one eval step for all workloads, without the real data iterator because it is not realistic to have all datasets available at testing time.

Adding an end-to-end test using MNIST to `submission_runner_test.py`.

Renaming `batch_size` to `global_batch_size` everywhere in the API so that we use it consistently.

Making both MNIST versions feature complete.

Cleaning up some WMT and other workloads.

### CODE REVIEW

File: algorithmic_efficiency/workloads/mnist/workload.py [hunk 1]
```diff
@@ -19,10 +26,18 @@ def loss_type(self):
 
   @property
   def num_train_examples(self):
-    return 60000
+    return 50000
 
   @property
-  def num_eval_examples(self):
+  def num_eval_train_examples(self):
+    return 10000
```
Comment (fsschneider) [inline algorithmic_efficiency/workloads/mnist/workload.py:33]: I may be mistaken here, but I think currently, we are using `num_eval_train_examples = num_train_examples = 50000`, not `10000`, no?
Judging from:

https://github.com/znado/algorithmic-efficiency/blob/17fc48fd0062289f05c9eb62d1d45f72e5837d08/algorithmic_efficiency/workloads/mnist/mnist_pytorch/workload.py#L62-L63

we are using the same dataset for both `train` and `train_eval`.

If I didn't make a mistake, the same would be true for the `ImageNet` workload as well.

File: algorithmic_efficiency/workloads/ogbg/ogbg_jax/workload.py [hunk 2]
```diff
@@ -176,27 +169,33 @@ def _eval_batch(self, params, graphs, labels, masks, model_state, rng):
...
+  def _eval_model_on_split(self,
+                           split: str,
+                           num_examples: int,
+                           global_batch_size: int,
+                           params: spec.ParameterContainer,
+                           model_state: spec.ModelAuxiliaryState,
+                           rng: spec.RandomState,
+                           data_dir: str) -> Dict[str, float]:
     """Run a full evaluation of the model."""
+    # DO NOT SUBMIT use num_examples
```
Comment (fsschneider) [inline algorithmic_efficiency/workloads/ogbg/ogbg_jax/workload.py:181]: Same questions as above.

File: algorithmic_efficiency/workloads/wmt/wmt_jax/workload.py [hunk 3]
```diff
@@ -475,28 +389,41 @@ def output_activation_fn(self,
...
+        lambda x: x / eval_denominator,  # pylint: disable=cell-var-from-loop
+        eval_metrics_sums)
 
-    _, bleu_score = self.translate_and_calculate_bleu(
-        p_pred_step=self._p_pred_step,
-        p_init_cache=self._p_init_cache,
+    bleu_score = self.translate_and_calculate_bleu(
         target=params,
-        predict_ds=self._predict_ds,
+        predict_ds=self._predict_ds, # DO NOT SUBMIT delete
```
Comment (fsschneider) [inline algorithmic_efficiency/workloads/wmt/wmt_jax/workload.py:424]: Is this comment still relevant?

**Review by fsschneider — CHANGES_REQUESTED** [review]
Overall LGTM! Thanks for the great PR which does a lot of important and great cleaning up.

I just had some minor comments and questions.

Two more things that I wanted to mention that didn't fit as inline comments:

1. Could you fix the `README.md` so that under `Running a workload` it uses the correct paths? I think the only changes needed are changing `python3 algorithmic_efficiency/submission_runner.py` to `python3 submission_runner.py` for both JAX and PyTorch. I did this in PR #65 but I will close this one as it is superseded by yours.
2. Could you make sure that the PR passes all the Github Actions, e.g. linting and testing?

Thanks again for this great PR!

File: algorithmic_efficiency/spec.py [hunk 4]
```diff
@@ -214,11 +218,59 @@ def loss_fn(
...
+        'validation',
+        num_examples=self.num_validation_examples,
+        global_batch_size=global_batch_size,
+        params=params,
+        model_state=model_state,
+        rng=rng,
+        data_dir=data_dir)
+    for k, v in validation_metrics.items():
+      eval_metrics['validation/' + k] = v
+    # Evaluate on the test set if we have one.
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/spec.py:259]: Maybe? I think we should figure out what to do for validation sets for all workloads, added it to the meeting agenda for this week.

File: algorithmic_efficiency/workloads/imagenet/imagenet_pytorch/workload.py [hunk 5]
```diff
@@ -32,31 +32,37 @@ def cycle(iterable):
...
+  def _eval_model_on_split(self,
+                           split: str,
+                           num_examples: int,
+                           global_batch_size: int,
+                           params: spec.ParameterContainer,
+                           model_state: spec.ModelAuxiliaryState,
+                           rng: spec.RandomState,
+                           data_dir: str):
     """Run a full evaluation of the model."""
+    # DO NOT SUBMIT use num_examples
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/imagenet/imagenet_pytorch/workload.py:54]: going to dismiss for now until we figure out any imagenet validation sets.

File: algorithmic_efficiency/workloads/ogbg/ogbg_jax/workload.py (hunk 2 above)
Comment (znado) [reply algorithmic_efficiency/workloads/ogbg/ogbg_jax/workload.py:181]: fixed

File: algorithmic_efficiency/spec.py [hunk 6]
```diff
@@ -214,11 +218,59 @@ def loss_fn(
...
+
   def eval_model(self,
+                 global_batch_size: int,
                  params: ParameterContainer,
                  model_state: ModelAuxiliaryState,
-                 rng: RandomState):
+                 rng: RandomState,
+                 data_dir: str) -> Dict[str, float]:
     """Run a full evaluation of the model."""
+    # DO NOT SUBMIT handle the case where batch size > num examples or when num_examples % batch_size != 0
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/spec.py:238]: yeah I think this is probably fine then, I just added this defensively to make sure I didn't leave it unchecked (I meant to follow up on it before sending the PR). thanks for checking!

**Review by fsschneider — APPROVED** [review]

**Comment by runame** [comment]
@znado When trying to merge the changes from this PR into my `wmt-pytorch` branch, I noticed that it seems to break the WMT Jax workload: you have [changed](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/wmt/wmt_jax/input_pipeline.py#L257-L265) the `get_wmt_datasets()` function to `get_wmt_dataset()` in the input pipeline, but in the workload the `build_input_queue()` function is [not adjusted accordingly](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/wmt/wmt_jax/workload.py#L308). I assume you want to call `build_input_queue()` for each dataset separately in the workload?

**Comment by znado** [comment]
> @znado When trying to merge the changes from this PR into my `wmt-pytorch` branch, I noticed that it seems to break the WMT Jax workload: you have [changed](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/wmt/wmt_jax/input_pipeline.py#L257-L265) the `get_wmt_datasets()` function to `get_wmt_dataset()` in the input pipeline, but in the workload the `build_input_queue()` function is [not adjusted accordingly](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/wmt/wmt_jax/workload.py#L308). I assume you want to call `build_input_queue()` for each dataset separately in the workload?

Ah yes, at the last minute I reverted some of the WMT changes to make in a more comprehensive follow-up PR, but must have only partially reverted them. Thanks for catching that, you're totally right. Feel free to revert/refactor anything as needed to get it running again, and I'll hopefully have the other PR(s) that continue cleaning stuff up done soon.

**Comment by runame** [comment]
> Ah yes, at the last minute I reverted some of the WMT changes to make in a more comprehensive follow-up PR, but must have only partially reverted them. Thanks for catching that, you're totally right. Feel free to revert/refactor anything as needed to get it running again, and I'll hopefully have the other PR(s) that continue cleaning stuff up done soon.

Thanks for the quick reply! I will just adjust the workload to the new `get_wmt_datasets()` function for now.
